### DISCUSSION TITLE
Conversion of raw values to final values

### DISCUSSION THREAD

**Discussion by gekkehenkie11** [emporia-vue-local/esphome#284] [General]
I think I'm pretty close to solving the apparent/real power issue. But I do need to know, how does the official ESP code exactly do its conversion? How did you guys come up with the conversion of raw power values to power values for example, @flaviut ? Did you reverse engineer the firmware? Or did you try to come up with values that makes sense?

One thing I don't really understand for example is that calibration value in the YAML file. Is this something that on stock firmware is the same for everybody (so in other words, it doesnt exist on stock firmware)? I know the atmel chip uses a calibration value for the measurements so that SHOULD mean that no calibration would even be necessary on the ESP side, since it's already being done by the Atmel chip? 

Anyway, I was hoping you could give me some pointers because to fix the apparent/real power issue I need the exact conversion calculation that the ESP uses. Thanks

**Comment by flaviut** [comment]
> How did you guys come up with the conversion of raw power values to power values for example, @flaviut ?

I didn't do it, but it was basically just linear interpolation: https://github.com/emporia-vue-local/emporia-vue2-reversing/issues/1#issuecomment-977171995

I did try and reverse engineer the firmware, but the ISA isn't fully supported by ghidra, and one of the missing things was floating points.

> One thing I don't really understand for example is that calibration value in the YAML file.

This does come from reverse engineering. [I noticed there was a table for this calibration value, and that the correct values depended on which voltage wire is connected to each phase. ](https://github.com/emporia-vue-local/emporia-vue2-reversing/issues/1#issuecomment-980827291)I wasn't able to figure out the exact logic

**Comment by gekkehenkie11** [comment]
Hmm that's interesting indeed. I just also checked from the log file it says:

[0;31mE (3859) calibration: Load calibration fail, using defaults
[0;32mI (3869) calibration: Set calibration, setDefault=1, defaultOverride=1, channel: 0, val: 0.022533599
[0;32mI (3879) calibration: Set calibration, setDefault=1, defaultOverride=1, channel: 1, val: 0.022000000
[0;32mI (3889) calibration: Set calibration, setDefault=1, defaultOverride=1, channel: 2, val: 0.022000000

So there indeed doesn't seem to be any factory calibration data per unit. At least not for the ESP. The Atmel chip does seem to load some calibdata though, not sure if that IS factory calibrated or not.

**Comment by gekkehenkie11** [comment]
I'm confused though. Maybe you understand this @flaviut: For one of the smaller CT's I get power reading 0x77A1B in the ESP sensor reading packet. So if I understand correctly you should calculate: 0x77A1B  * 0.022 /22 = 490 Watt. However it shows (correctly) something like 360 Watt in the app (using all stock firmware). Any idea what I'm doing wrong?

**Reply by flaviut** [reply]
That calculation seems right, and it matches the spreadsheet: https://docs.google.com/spreadsheets/d/1Yu98tDRhFJqqMntNak6X297SKyzQtf0_Cvv_g7qKewk/edit#gid=0

I'm not sure what's wrong here.

**Reply by gekkehenkie11** [reply]
hmm weird. I also just tried to start the mqtt server on 'emporia' wifi SSID like you described in your article. However it seems once you have had it configured to your home wifi, it never looks for 'emporia' wifi again !?!? It keeps rescanning for my home wifi ...

**Reply by gekkehenkie11** [reply]
@flaviut I just tested it with your firmware, it says indeed 490 Watt. However that's not correct :) I hooked up another VUE2 with stock firmware and it shows 360 watt, which IS correct. Not sure what's going on here but I'm starting to think there's more to the conversion than the calculation above ... I mean it MUST be, since the stock firmware shows quite a different result based on the same data....

Anyway, like noted above I've been trying to set up that MQTT server at the "esphome" ssid. But it seems it's not even trying to connect to it, it keeps looking for my home network, to which I of course had programmed it before. Do you maybe have an idea? I would love to double check the raw sensor data that way to confirm the above.

**Reply by flaviut** [reply]
Not sure about the numbers, but the SSID & MQTT stuff [gets saved in the ESP32 NVS](https://flaviutamas.com/2021/reversing-emporia-vue-2#non-volatile-storage). You can dump your NVS & compare it [to a clean version](https://github.com/emporia-vue-local/emporia-vue2-reversing/blob/master/nvs.json), then re-flash it.

**Comment by gekkehenkie11** [comment]
Ok I still had a virgin VUE2 lying around, tested it and this time I started receiving MQTT messages. So it seems it only works for virgin VUE2's. Not sure how to put a VUE2 back into virgin state. Anyway, now everything looked fine, got 0x5F8EC as value which yields 393 Watt which seems fine ... Really not sure what's going on, just tested again with my 'development' board and got those weird values again, with stock firmware Atmel. Starting to think I might have messed up my board. OR there's a different atmel firmware on virgin boards ... Only one way to find out :)

**Comment by gekkehenkie11** [comment]
Ok yeah did some more testing, the other board is still fine, it's just that if I hook up even a single pin to the JLINK, the measurements mess up. So I guess the only way to get what I want (sample buffers of A and V) is to send them via I2C to the ESP and send them to HA. Damn that's going to cost me some time to fix that all up ...

*EDIT* hmm probably easiest to dump results to the UART
*EDIT* actually same thing, I hook up the UART and all readings mess up again. Did you ever notice that @flaviut ?
